flagext/stringslicecsv.go

```diff
@@ -13,7 +13,9 @@ func (v StringSliceCSV) String() string {
 
 // Set implements flag.Value
 func (v *StringSliceCSV) Set(s string) error {
-	*v = strings.Split(s, ",")
+	if s != "\"\"" && len(s) > 0 {
```

Comment:
This feel a wrong fix to me and I would like to learn more about it: who calls `Set()` with the string content set to `""`?

Reply:
It is a bit strange, `Set` is called in `UnmarshalYAML` when unmarshalling the flag string. We also call `Set` in tests. The added test calls Set, Marshal, and Unmarshal, by doing the check here in `Set` we won't fill the pointer to the `[]String` in the struct, which when marshaled is seen as a zero value and correctly marshaled to `""` rather than `'""'`

Reply:
Let's skip the usage in tests. What I don't understand is what's the real use case where `UnmarshalYAML()` calls `Set()` passing a string with content `""`.

Reply:
I see it originates from a failure you're seeing in https://github.com/grafana/mimir/pull/2583.

Reply:
This issue looks weird to me. If you add the following test to `runtimeconfig/manager_test.go` it passes:

```
func Test_Config_Unmarshal(t *testing.T) {
	type TestStruct struct {
		Config Config `yaml:"runtime_config"`
	}

	data := `
runtime_config:
    period: 10s
    file: ""
`

	var testStruct TestStruct
	require.NoError(t, yaml.Unmarshal([]byte(data), &testStruct))
	require.Equal(t, 10*time.Second, testStruct.Config.ReloadPeriod)
	require.Equal(t, "", testStruct.Config.LoadPath.String())
}
```

I've the feeling dskit is fine, and we're doing something wrong in Mimir.

Reply:
The zero value problem isn't just an issue with unmarshaling, it's an issue with marshaling and unmarshaling. 

The test that is failing in Mimir `TestConfigAPIEndpoint` follows the same pattern as the test added in dskit. Where we load a config into a Mimir service without a runtime config, we call the `/config` endpoint to dump the config and start a new Mimir service with the config from that API endpoint. When we marshal the config we set the `-runtime-config.file` in the yaml output to `""`, which is correct as it is unset, but when we go to unmarshal we attempt to split/manipulate the string `""` causing the runtime file to be read in as `/"/"` and we will attempt to load a file named `\"\"` and this change is to prevent that from happening by making sure we handle the string `""` as a zero string that is unset/unused as yaml would for a normal string.

Reply:
I finally understood the problem. You can reproduce it with this test, that doesn't involve calling `Set("\"\"")` (as [done here](https://github.com/grafana/dskit/pull/206/files#diff-ad046aeabdc7ca6dcd1bee4444bf9e9e5479af21a8089a96ab905da65aace996R42-R43)):

```
func Test_StringSliceCSV_MarshalAndThenUnmarshalEmptyFile(t *testing.T) {
	type TestStruct struct {
		CSV StringSliceCSV `yaml:"csv"`
	}

	original := TestStruct{
		CSV: nil,
	}

	data, err := yaml.Marshal(&original)
	require.NoError(t, err)

	// Now unmarshal the YAML from the marshalled one.
	var unmarshalled TestStruct
	require.NoError(t, yaml.Unmarshal(data, &unmarshalled))
	require.Equal(t, original, unmarshalled)
	require.Equal(t, StringSliceCSV(nil), unmarshalled.CSV)
}
```

The actual problem is in the `strings.Split(s, ",")` behaviour. If you pass an empty string to `strings.Split()` it doesn't return an empty slice, but a slice with an empty string.

I think the proper fix is changing the `Set()` function as follows:

```
func (v *StringSliceCSV) Set(s string) error {
	if len(s) == 0 {
		*v = nil
		return nil
	}

	*v = strings.Split(s, ",")
	return nil
}
```

Reply:
```
original := TestStruct{
		CSV: nil,
	}
```
That's not quite the same as the marshaled config from the config endpoint. The config we receive from the config API is: 
```
...
runtime_config:
    period: 10s
    file: ""
...
```
Meaning our supplied string to `Set()` after the unmarshal would be `""` rather than a nil pointer. The problem doesn't happen on the first creation of the stringsplitcsv but rather when we go to read in and unmarshal a config.

Reply:
I don't think what you're saying is correct. For example, if in the test I suggested in the previous comment you print out  `fmt.Println(string(data))` after the call to `data, err := yaml.Marshal(&original)` you will get:

```
csv: ""
```

In YAML, the entry `csv: ""` means that the field `csv` is **an empty string** and not a string with content `""`.

As explained in my previous comment, `Set()` will not be called with a string whose content is `""`, but `Set()` will be called with an empty string and the actual problem is the call to `strings.Split()` because it returns `[""]` if you pass an empty string in input.

Reply:
Yeah I think I see what you mean here, I updated the test and logic in the PR to reflect this.